src/Elf.cpp

```diff
@@ -0,0 +1,989 @@
+            symbols[plt_def] = plt_def;
+        }
+    }
+
+    // Start placing the sections into the shared object.
+
+    // Leave room for the header, and program headers at the beginning of the file.
+    append_zeros(output, sizeof(ehdr));
+    append_zeros(output, sizeof(phdrs[0])*3);
+
+    // We need to perform the relocations. To do that, we need to position the sections
+    // where they will go in the final shared object.
+    write_section(plt);
+    for (const Section &s : obj.sections()) {
+        if (s.is_alloc() && !s.is_writable()) {
+            write_section(s);
+        }
+    }
+    append_padding(output, 4096);
+    text_phdr.p_filesz = output.size() - text_phdr.p_offset;
+
+    data_phdr.p_type = PT_LOAD;
+    data_phdr.p_flags = PF_W | PF_R;
+    safe_assign(data_phdr.p_offset, output.size());
+    data_phdr.p_align = 4096;
+    for (const Section &s : obj.sections()) {
+        if (s.is_alloc() && s.is_writable()) {
+            write_section(s);
+        }
+    }
+
+    // The got will be written again later, after we add entries to it.
+    write_section(got);
+
+    /// Now that we've written the sections that define symbols, we
+    // can generate the symbol table.
+    Section symtab(".symtab", Section::SHT_SYMTAB);
+    symtab.set_alignment(4);
+    symtab.set_flag(Section::SHF_ALLOC);
+    std::vector<Sym<T>> syms;
+    Sym<T> undef_sym;
+    memset(&undef_sym, 0, sizeof(undef_sym));
+    syms.push_back(undef_sym);
+    // TODO: This map with pointers as the key will lead to non-deterministic builds...
+    std::map<const Symbol *, uint16_t> symbol_idxs;
+    uint64_t local_count = 0;
+    for (bool is_local : {true, false}) {
+        for (const auto &i : symbols) {
+            const Symbol *s = i.second;
+            if ((s->get_binding() == Symbol::STB_LOCAL) != is_local) continue;
+
+            uint64_t value = s->get_offset();
+            // In shared objects, the symbol value is a virtual address,
+            // not a section offset.
+            if (s->is_defined()) {
+                value += get_section_offset(*s->get_section());
+            }
+            Sym<T> sym;
+            safe_assign(sym.st_name, strings.get(s->get_name()));
+            safe_assign(sym.st_value, value);
+            safe_assign(sym.st_size, s->get_size());
+            sym.set_type(s->get_type());
+            sym.set_binding(s->get_binding());
+            safe_assign(sym.st_other, s->get_visibility());
+            sym.st_shndx = section_idxs[s->get_section()];
+
+            safe_assign(symbol_idxs[s], syms.size());
+            syms.push_back(sym);
+        }
+        if (is_local) {
+            local_count = syms.size();
+        }
+    }
+    symtab.set_contents(syms);
+    uint16_t symtab_idx = write_section(symtab, sizeof(syms[0]));
+    safe_assign(shdrs[symtab_idx].sh_info, local_count);
+
+    // Also write the symbol table as SHT_DYNSYM.
+    Section dynsym = symtab;
```

Comment:
For shared objects, the symbol table is useless, you can save on memory space by just creating dynsym.